unit/ctest_array_average.c

```diff
@@ -0,0 +1,1169 @@
+
+  struct gkyl_rect_grid grid;
+  gkyl_rect_grid_init(&grid, ndim, lower, upper, cells);
+
+  struct gkyl_basis basis;
+  gkyl_cart_modal_serendip(&basis, ndim, poly_order);
+
+  int ghost[] = {1};
+  struct gkyl_range local, local_ext;
+  gkyl_create_grid_ranges(&grid, ghost, &local_ext, &local);
+
+  // define the reduced range and basis for averaging
+  struct gkyl_range red_local, red_local_ext;
+  gkyl_range_init(&red_local, 1, &local.lower[0], &local.lower[0]);
+  gkyl_range_init(&red_local_ext, 1, &local_ext.lower[0], &local_ext.lower[0]);
+  struct gkyl_basis red_basis;
+  gkyl_cart_modal_serendip(&red_basis, 1, poly_order);
+
+  // project the target function and weight
+  gkyl_proj_on_basis *projf = gkyl_proj_on_basis_new(
+      &grid, &basis, poly_order + 1, 1, evalFunc_1x, NULL);
+
+  struct gkyl_array *fx_c = mkarr(basis.num_basis, local_ext.volume, use_gpu);
+  struct gkyl_array *fx_c_ho = use_gpu? mkarr(fx_c->ncomp, fx_c->size, false) : gkyl_array_acquire(fx_c);
+  gkyl_proj_on_basis_advance(projf, 0.0, &local, fx_c_ho);
+  gkyl_array_copy(fx_c, fx_c_ho);
+
+  gkyl_proj_on_basis_release(projf);
+
+  gkyl_proj_on_basis *projw = gkyl_proj_on_basis_new(
+      &grid, &basis, poly_order + 1, 1, evalWeight_1x, NULL);
+
+  struct gkyl_array *wx_c = mkarr(basis.num_basis, local_ext.volume, use_gpu);
+  struct gkyl_array *wx_c_ho = use_gpu? mkarr(wx_c->ncomp, wx_c->size, false) : gkyl_array_acquire(wx_c);
+  gkyl_proj_on_basis_advance(projw, 0.0, &local_ext, wx_c_ho);
+  gkyl_array_copy(wx_c, wx_c_ho);
+
+  gkyl_proj_on_basis_release(projw);
+
+  // compute weighted average
+    int avg_dim_x[] = {1,0,0};
+    struct gkyl_array_average_inp inp_avg_full = {
+    .grid = &grid,
+    .basis = basis,
+    .basis_avg = red_basis,
+    .local = &local,
+    .local_avg = &red_local,
+    .local_avg_ext = &red_local_ext,
+    .weight = wx_c,
+    .avg_dim = avg_dim_x,
+    .use_gpu = use_gpu
+  };
+  struct gkyl_array_average *avg_full = gkyl_array_average_new(&inp_avg_full);
+
+  struct gkyl_array *avgf_c = mkarr(red_basis.num_basis, red_local.volume, use_gpu);
+  gkyl_array_average_advance(avg_full, fx_c, avgf_c);
+
+  gkyl_array_average_release(avg_full);
+
+  // fetch and transfer results
+  double * avg_c0 = gkyl_array_fetch(avgf_c, 0);
+  double *avg_c0_ho = gkyl_malloc(sizeof(double));
+  if (use_gpu)
+    gkyl_cu_memcpy(avg_c0_ho, avg_c0, sizeof(double), GKYL_CU_MEMCPY_D2H);
+  else
+    memcpy(avg_c0_ho, avg_c0, sizeof(double));
+
+  // check results
+  double intf_ref = solution_array_integrate(grid, basis, local_ext, local, wx_c, fx_c, use_gpu);
+  double intw_ref = solution_array_integrate(grid, basis, local_ext, local, NULL, wx_c, use_gpu);
+  double solution = intf_ref/intw_ref;
+  double result = avg_c0_ho[0]*0.5*sqrt(2);
+  double rel_err = fabs(result - solution) / fabs(solution);
+  printf("Result: %g, solution: %g\n",result,solution);
+  printf("Relative error: %e\n", fabs(result - solution) / fabs(solution));
```

Comment:
comment these print statements out. They are useful during dev and debugging, but once the unit test is in place one only want to know if it passes or fails, and not receive numbers on the screen that only make sense to the author.